protocols/relay/src/v2/copy_future.rs

```diff
@@ -0,0 +1,149 @@
+use std::convert::TryInto;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+pub struct CopyFuture<S, D> {
+    src: BufReader<S>,
+    dst: BufReader<D>,
+
+    max_circuit_duration: Delay,
+    max_circuit_bytes: u64,
+    bytes_sent: u64,
+}
+
+impl<S: AsyncRead, D: AsyncRead> CopyFuture<S, D> {
+    pub fn new(src: S, dst: D, max_circuit_duration: Duration, max_circuit_bytes: u64) -> Self {
+        CopyFuture {
+            src: BufReader::new(src),
+            dst: BufReader::new(dst),
+            max_circuit_duration: Delay::new(max_circuit_duration),
+            max_circuit_bytes,
+            bytes_sent: Default::default(),
+        }
+    }
+}
+
+impl<S, D> Future for CopyFuture<S, D>
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+    D: AsyncRead + AsyncWrite + Unpin,
+{
+    type Output = io::Result<()>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = &mut *self;
+
+        loop {
+            if this.bytes_sent > this.max_circuit_bytes {
+                return Poll::Ready(Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Max circuit bytes reached.",
+                )));
+            }
+
+            enum Status {
+                Pending,
+                Done,
+                Progressed,
+            }
+
+            let src_status = match forward_data(&mut this.src, &mut this.dst, cx) {
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+                Poll::Ready(Ok(0)) => Status::Done,
+                Poll::Ready(Ok(i)) => {
+                    this.bytes_sent += i;
+                    Status::Progressed
+                }
+                Poll::Pending => Status::Pending,
+            };
+
+            let dst_status = match forward_data(&mut this.dst, &mut this.src, cx) {
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+                Poll::Ready(Ok(0)) => Status::Done,
+                Poll::Ready(Ok(i)) => {
+                    this.bytes_sent += i;
+                    Status::Progressed
+                }
+                Poll::Pending => Status::Pending,
+            };
+
+            match (src_status, dst_status) {
+                // Both source and destination are done sending data.
+                (Status::Done, Status::Done) => return Poll::Ready(Ok(())),
+                // Either source or destination made progress.
+                (Status::Progressed, _) | (_, Status::Progressed) => {}
+                // Both are pending. Check if max circuit duration timer fired, otherwise return
+                // Poll::Pending.
+                (Status::Pending, Status::Pending) => break,
+                // One is done sending data, the other is pending. Check if timer fired, otherwise
+                // return Poll::Pending.
+                (Status::Pending, Status::Done) | (Status::Done, Status::Pending) => break,
```

Comment:
True, though the reasons for the two code paths differ and thus the comments do as well.

I could write the following, though I am not sure that is more intuitive:

``` rust
                // Both are pending. Check if max circuit duration timer fired, otherwise return
                // Poll::Pending.
                (Status::Pending, Status::Pending) |
                // One is done sending data, the other is pending. Check if timer fired, otherwise
                // return Poll::Pending.
                (Status::Pending, Status::Done) | (Status::Done, Status::Pending) => break,
```

Preference or alternative idea?

Reply:
On second thought, I think this is okay as it.